File: Swift/프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.swift
import Foundation

fileprivate class Stack {
    private var stack: [Int] = []
    
    var isEmpty: Bool {
        return stack.isEmpty
    }
    
    var count: Int {
        return stack.count
    }
    
    var peek: Int? {
        return stack.last
    }
    
    func pop() -> Int? {
        return stack.popLast()
    }
    
    func push(_ element: Int) {
        stack.append(element)
    }
}

func solution(_ board:[[Int]], _ moves:[Int]) -> Int {
    let n: Int = board.count
    var result: Int = 0
    
    var columnStack: [Stack] = (0..<n).map { _ in Stack() }
    var stack: Stack = Stack()
    
    for column in 0..<n {
        let rStack = columnStack[column]
        for row in 0..<n {
            let item = board[n-1-row][column]
            guard item != 0 else { continue }
            rStack.push(item)
        }
    }
    
    for move in moves {
        let moveIndex: Int = move - 1
        guard let item = columnStack[moveIndex].pop() else { continue }
        if !stack.isEmpty && (stack.peek == item) {
            stack.pop()
            result += 2
        } else {
            stack.push(item)
        }
    }
    

    return result
}

File: Java/프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.Stack;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        int n = board.length;
        Stack<Integer>[] items = new Stack[n];
        for (int i = 0; i < n; i++) {
            items[i] = new Stack<Integer>();
        }
        
        Stack<Integer> resultStack = new Stack<>();
        
        for (int y = n-1; y >= 0; y--) {
            for(int x = 0; x < n; x++) {
                int item = board[y][x];
                if (item != 0) {
                    items[x].push(item);
                }
            }
        }
        
        for (int move : moves) {
            int moveIndex = move - 1;
            if (!items[moveIndex].isEmpty()) {
                int pickedItem = items[moveIndex].pop();
                if (!resultStack.isEmpty() && pickedItem == resultStack.peek()) {
                    resultStack.pop();
                    answer += 2;
                } else {
                    resultStack.push(pickedItem);
                }
            } else {
                continue;
            }
        }
        return answer;
    }
}